src/finch/finch_logic/nodes.py

```diff
@@ -1,7 +1,33 @@
+from __future__ import annotations
+
 from abc import abstractmethod
+from collections.abc import Callable
 from dataclasses import dataclass
-from typing import Any, Iterable
-from ..symbolic import Term
+from typing import Any, Never, Self, TypeVar
+
+from ..symbolic import Expression, Leaf, Term
```

Comment:
Symbolic should stay as is, let's handle this with is_expr rather than types.

Reply:
One reason to use is_expr is that we may opt for IR nodes which are all the same python class, represented by an integer enum, for example.